### DISCUSSION TITLE
Daily note template based on day of the week (bonus: link note to current day)

### DISCUSSION THREAD

**Discussion by deathau** [SilentVoid13/Templater#240] [Templates Showcase]
I wanted a solution to populate different Day Planner schedules based on the current day of the week, and with some help from @shabegom I have achieved this with the following:

### scheduleTemplate.js
```javascript
function scheduleTemplate(tp) {
  switch (parseInt(moment(tp.file.title).format('d'))) {
    case 1: return tp.file.include("[[Schedule - 01 Monday]]")
    case 2: return tp.file.include("[[Schedule - 02 Tuesday]]")
    case 3: return tp.file.include("[[Schedule - 03 Wednesday]]")
    case 4: return tp.file.include("[[Schedule - 04 Thrusday]]")
    case 5: return tp.file.include("[[Schedule - 05 Friday]]")
    case 6:
    case 0: return tp.file.include("[[Schedule - Weekend]]")
    default: return tp.file.include("[[Schedule - +Weekday]]")
  }
}

module.exports = scheduleTemplate
```
### command.js
```javascript
function command(action) {
  const allCommands = app.commands.listCommands();
  const command = allCommands.filter(
    (command) => command.name.toUpperCase() === action.toUpperCase().trim()
  )[0];
  app.commands.executeCommandById(command.id);
  return '';
};

module.exports = command;
```

## Daily note Template
```
<% tp.user.scheduleTemplate(tp) %>
<%* if(moment(tp.file.title).isSame(new Date(), "day")) tp.user.command("Day Planner: Link today's Day Planner to the current note") %>
```

**Comment by megclaypool** [comment]
I was *struggling* to figure out how to include partial templates in other templates -- this is exactly what I was trying to accomplish! And linking the day planner to today's note is the icing on the cake! 🎉

Thanks!!! ❤️

**Comment by help-moo** [comment]
Hi everyone, I m really interested in using this feature but i really don't know how. Can u please helpe me out. 
1. Do i have to create two .js files: scheduleTemplate and command ?
2. Where do i put them? I assume i have to link that to the script files folder ( in the Plugins option) 
3. What info i put in "function name" and "system command" ( in the Plugins options)
4. Do i need to change anything in the script because i use Obsidian in spanish? 
5. As far as i know i will need to create templates named after the days i want to be modified? how should i name those templates ? 

i know im such a n00b but i really want to use this. 

thnks a lot 

- moooooo

**Reply by megclaypool** [reply]
I'll jump in, since I was planning on writing a how-to anyhow (so future-me doesn't forget how I set this all up 😆)

- You need to create a folder for scripts someplace inside your vault. 
  - I just put mine in the root, and called it TemplaterJS, but you can put it anywhere inside your vault and call it whatever you like.  
  - You should also have a folder for templates, but you've probably already got that set up 😀 
- In the Templater preferences, put the path (from your vault root) to your scripts folder in the "Script files folder location"
- You don't need to turn on the "Enable System Commands" or put anything in the function name or system commands boxes
- Using a text editor, create a file called scheduleTemplate.js in the scripts folder you just created. Paste in the contents you see above. Do the same for command.js
  - It's really important to use a text editor, like NotePad or TextEdit (or even better, VSCode or Atom if you've already got one of them) rather than a word processor like Word or Pages. 
- If you look at the contents of the scheduleTemplate.js, you'll see a line that says `switch (parseInt(moment(tp.file.title).format('d'))) {`. What that's doing is telling the program to look at the file title and treat it like a date, figure out which day of the week it is, and return a number from 0 - 6 (so this trick only works on notes whose name is based on a date, like daily notes). Then it has a separate case for each number, and includes a file for each day. 
  - Depending on where you live, either Sunday or Monday may be 0, so you may need to change which templates load for which case
  - You can also change the names of the templates being included. The script is using the file title (or more specifically, what you would put in double wiki brackets to link to that file, which gets a little more complicated if there are files with duplicate names 🙄 😛 )
  - You can put the same template for more than one day. (When I was a student, I had the same schedule on M/W/F and on T/Th)
- In your templates folder, create a file (just in Obsidian -- you don't need a text editor for this 😀) for each different schedule you want to embed
  - In each of these schedule templates, put whatever custom bits you want to have for that/those specific day(s)
- Also create a file for your daily notes template, and into that put all the stuff you want to appear on *every* day. Wherever you want the stuff that's specific for each day, paste the `<% tp.user.scheduleTemplate(tp) %>
<%* if(moment(tp.file.title).isSame(new Date(), "day")) tp.user.command("Day Planner: Link today's Day Planner to the current note") %>` code.

For example, here's my Daily Notes template (I'm also using the Tasks plugin, which is what all the fol-de-rol at the bottom is about):
```
# Daily Log -- {{ date:MMM D, YYYY}}

[[<% tp.date.now("YYYY[ -- Year]", 0, tp.file.title) %> |<% tp.date.now("YYYY", 0, tp.file.title) %>]] » [[<% tp.date.now("YYYY-MM", 0, tp.file.title) %> -- Month#<% tp.date.now("YYYYMMDD dddd, MMM D, YYYY", 0, tp.file.title) %> | <% tp.date.now("MMMM", 0, tp.file.title) %>]]  » [[<% tp.date.weekday("YYYY-MM-DD", 0, tp.file.title) %> -- Week#<% tp.date.now("YYYYMMDD dddd, MMM D, YYYY", 0, tp.file.title) %> | Week <% tp.date.now("ww", 0, tp.file.title) %>]]
[[<% tp.date.now("YYYYMMDD", -1, tp.file.title) %> | ⇦ Previous Day]] | [[<% tp.date.now("YYYYMMDD", +1, tp.file.title) %> | Next Day ⇨]] 

<% tp.user.daily_schedule(tp) %>
<%* if(moment(tp.file.title).isSame(new Date(), "day")) tp.user.command("Day Planner: Link today's Day Planner to the current note") %>
---

## Tasks
### Overdue 
- Personal
	```tasks
	not done
	path includes Task List
	due before {{ date:YYYY-MM-DD }}
	```

- Work
	```tasks
	not done
	path includes Work
	due before {{ date:YYYY-MM-DD }}
	```

### Due today
- Personal
	```tasks
	not done
	path includes Task List
	due on {{ date:YYYY-MM-DD }}
	```

- Work
	```tasks
	not done
	path includes Work
	due on {{ date:YYYY-MM-D D}}
	```

### Due w/in 7 days
- Personal
	```tasks
	not done
	path includes Task List
	heading includes running
	due after {{ date:YYYY-MM-DD }}
	due before {{ date+7d:YYYY-MM-DD }}
	```

- Work
	```tasks
	not done
	path includes Work
	due after {{ date:YYYY-MM-DD }}
	due before {{ date+7d:YYYY-MM-DD }}
	```

### No due date
- Personal
	```tasks
	not done
	path includes Task List
	no due date
	```

- Work
	```tasks
	not done
	path includes Work
	no due date
	```

### Done today
- Personal
	```tasks
	path includes Task List
	done on {{ date:YYYY-MM-DD }}
	```

- Work
	```tasks
	path includes Work
	done on {{ date:YYYY-MM-DD}}
	```

---
## Notes

- Personal
	- 


- Work
	-  
```
You can see that under the breadcrumb, I'm using @deathau's script to load my daily schedule. Here's an example of my Friday schedule:
```
---
## Day Planner
- [ ] 06:00 Get up, brush teeth, make tea, etc
- [ ] 06:30 Prep on computer
	- [ ] check email
	- [ ] preview day 
		- [ ] check calendar and weekly and monthly pages
		- [ ] what's for dinner? does anything need thawed?
	- [ ] protein bar breakfast of champions
- [ ] 07:00 Professional development
	- [The Complete JavaScript Course 2021: From Zero to Expert!](https://www.udemy.com/course/the-complete-javascript-course/)
	- [[The Complete JavaScript Course 2021]]
- [ ] 08:00 Work
- [ ] 11:30 Lunch 
- [ ] 12:00 Nap
- [ ] 12:30 BREAK
- [ ] 13:00 [[Dev Meeting]]
- [ ] 14:00 Work
- [ ] 17:00 Cook dinner
- [ ] 18:00 Eat dinner
- [ ] 18:30 Clean up
- [ ] 19:00 Practice Guitar
- [ ] 19:30 Kid 1's Bedtime
- [ ] 20:00 Free Time / Fallback Guitar Practice
- [ ] 20:30 Kid 2's Bedtime
- [ ] 21:15 Work out
- [ ] 21:45 Shower
- [ ] 22:30 END
```

The super awesome thing about @deathau's script is that if you create your daily note on that day, it automatically links it to the Day Planner plugin, which is way spiffy! 😀

**Reply by help-moo** [reply]
Woww, Thank you so much. I just made it! it took me some time but i did it. Im so happy. Your answer also help me to understand better the basics of templater.
Thanks a lot!

**Reply by jbguberman** [reply]
@megclaypool Thank you so much for writing up this guide and, in particular, for sharing how you're incorporating the Tasks plugin into this workflow! I had been trying to to use Todoist-Sync on my Daily Notes pages, filtering for a given page's date using `tp.file.title`, but that plugin wasn't able to parse Templater syntax or variables. Instead of hacking the Todoist plugin, it looks like I'll be switching over to Tasks.

In case this is helpful to anyone else reading this, after following these instructions, Templater didn't seem to actually be doing anything with the Templater syntax. Make sure "Trigger Templater on new file creation" in the Templater settings is toggled on - this solved the issue for me!

**Reply by jroachgit** [reply]
@megclaypool  Hey - thanks so much for the write up.
I'm keen to understand, how are your dates in your task queries dynamic to the 'daily note's date'?

`- Work
	```tasks
	not done
	path includes Work
	due before {{ date:YYYY-MM-DD }}
	````

**Comment by megclaypool** [comment]
@jroachgit , you caught me!😆  I use the daily notes/periodic notes plugins, and that date code was leftover from their date templating language. (I'll admit that it's a little more succinct, albeit more limited...) I've since changed it to use Templater code instead, for consistency, but back when I posted this my daily/weekly/monthly templates were kind of a mishmash 😛 

Here's what my daily template looks like now: 
```
# Daily Log -- <% tp.date.now("dddd, MMMM D, YYYY", 0, tp.file.title) %>

[[<% tp.date.now("YYYY[ -- Year]", 0, tp.file.title) %> |<% tp.date.now("YYYY", 0, tp.file.title) %>]] » [[<% tp.date.now("YYYY-MM", 0, tp.file.title) %> -- Month#<% tp.date.now("YYYYMMDD dddd, MMM D, YYYY", 0, tp.file.title) %> | <% tp.date.now("MMMM", 0, tp.file.title) %>]]  » [[<% tp.date.weekday("YYYY-MM-DD", 0, tp.file.title) %> -- Week#<% tp.date.now("YYYYMMDD dddd, MMM D, YYYY", 0, tp.file.title) %> | Week <% tp.date.now("ww", 0, tp.file.title) %>]]
[[<% tp.date.now("YYYYMMDD", -1, tp.file.title) %> | ⇦ Previous Day]] | [[<% tp.date.now("YYYYMMDD", +1, tp.file.title) %> | Next Day ⇨]] 

<% tp.user.daily_schedule(tp) %>
<%* if(moment(tp.file.title).isSame(new Date(), "day")) tp.user.command("Day Planner: Link today's Day Planner to the current note") %>
---

## Tasks
### Overdue 
- Personal
	```tasks
	not done
	path includes Task List
	due before <% tp.date.now('YYYY-MM-DD', 0, tp.file.title) %>
	```

- Work
	```tasks
	not done
	path includes Work
	due before <% tp.date.now('YYYY-MM-DD', 0, tp.file.title) %>
	```

### Due today
- Personal
	```tasks
	not done
	path includes Task List
	due on <% tp.date.now('YYYY-MM-DD', 0, tp.file.title) %>
	```

- Work
	```tasks
	not done
	path includes Work
	due on <% tp.date.now('YYYY-MM-DD', 0, tp.file.title) %>
	```

### Due w/in 7 days
- Personal
	```tasks
	not done
	path includes Task List
	due after <% tp.date.now('YYYY-MM-DD', 0, tp.file.title) %>
	due before <% tp.date.now('YYYY-MM-DD', "P+1W", tp.file.title) %>
	```

- Work
	```tasks
	not done
	path includes Work
	due after <% tp.date.now('YYYY-MM-DD', 0, tp.file.title) %>
	due before <% tp.date.now('YYYY-MM-DD', "P+1W", tp.file.title) %>
	```

### No due date
- Personal
	```tasks
	not done
	path includes Task List
	no due date
	```

- Work
	```tasks
	not done
	path includes Work
	no due date
	```

### Done today
- Personal
	```tasks
	path includes Task List
	done on <% tp.date.now('YYYY-MM-DD', 0, tp.file.title) %>
	```

- Work
	```tasks
	path includes Work
	done on <% tp.date.now('YYYY-MM-DD', 0, tp.file.title) %>
	```

---
## Notes

- Personal
	- 


- Work
	-  
```

**Comment by rapatel0** [comment]
Slight tweak of the function but allows customization of notes within obsidian. This way you can modify the included files within obsidian including using [[]] to autocomplete


```javascript
function scheduleWeekday(tp, files) {
  switch (parseInt(moment(tp.file.title).format('d'))) {

    case 1:
    case 2:
    case 3:
    case 4:
    case 5:
      return tp.file.include(files.weekday)

    case 0:
    case 6: return tp.file.include(files.weekend)
  }

}
module.exports = scheduleWeekday;

```

Invocation:
```javascript
<% 
    tp.user.scheduleDaily(tp, {
          weekday:  "[[Utilities/Templater/Templates/Blocks/Process/P-Weekdays]]",
          weekend:      "[[P-Weekends]]",
    }) 
%>

```

**Reply by hieu6** [reply]
hello, thank you for these! they look very clean. 

> Slight tweak of the function but **_allows customization of notes within obsidian_**. This way you can **_modify the included files within obsidian_** including using [[]] to autocomplete

would you mind re-explaining what you mean by this? how do your changes to op's scripts change how they could be used? learning, learning 😌 🙇🏻‍♂️

**Reply by megclaypool** [reply]
This is great! 😀 

In his scheduleWeekday function definition (in the javascript file), @rapatel0 has two parameters telling the function to expect both the tp object and something called "files". Later in the function, you can see that the function is expecting to find both "weekday" and "weekend" variables inside files, so it looks like files is an object.  He's set up his switch so that if it's a weekday, files.weekday gets called and if it's a weekend, files.weekend gets called. (You could do a switch statement that calls a different template for each day of the week -- you'd just need to set up files.monday, files.tuesday, etc if that's what you wanted to do.)

Then in his Obsidian template, when he calls the scheduleDaily function (I think there was a bit of a typo here, and it was supposed to be scheduleWeekday), he passes in two arguments; tp *and* a mysterious object (which he creates right there -- this is that 2nd mysterious parameter, files, from his function declaration!) which contains the keys "weekday" and "weekend" and whose values are links to his weekday and weekend templates. (You notice that the link to the weekday template is much longer? That's because he must have another file somewhere in his vault that's also named "P-Weekdays", so his link for that file included the path to be more specific. I'm pretty sure autocomplete handled that automagically 😀) 

There are a couple of big advantages to doing it this way, and I think I might switch! First, because you're writing the part of the code containing the links *in Obsidian*, when you're originally creating the object and you start typing your links, `[[THE_TITLE_OF_THE_NOTE_YOU_WANT_TO_USE...` will autocomplete. That's super awesome cool right there. Secondly, I *think* that if you change the title or location of the note you're linking to, the link will auto-update, which is also magically delicious!

**Reply by rapatel0** [reply]
Yeah Exactly. 

The link auto updates if you move it in obsidian. 

It's also easy to tweak. Just add return values or  case statements for additional keys. Then add associated entries to the input JSON

Example: - ( If we wanted to add a monday note)

```javascript
function scheduleWeekday(tp, files) {
  switch (parseInt(moment(tp.file.title).format('d'))) {

    case 1:
      return tp.file.include(files.monday)
    case 2:
    case 3:
    case 4:
    case 5:
      return tp.file.include(files.weekday)

    case 0:
    case 6: return tp.file.include(files.weekend)
  }

}
```
```javascript
<% 
    tp.user.scheduleDaily(tp, {
          Monday:  "[[P-Monday]]",
          weekday:  "[[Utilities/Templater/Templates/Blocks/Process/P-Weekdays]]",
          weekend:      "[[P-Weekends]]",
    }) 
%>
```

**Comment by Nudelsalad** [comment]
Hey,
the linking of days works excellent only the day-planner seems not to work anymore in Obsidian 1.6.7 and day-planner 0.22.1
It seems like the current version Day-planner doesnt support the command "Day Planner: Link today's Day Planner to the current note" anymore. 
Listing all commands i only see these for the Day planner:

```json
  {
        "id": "obsidian-day-planner:re-sync",
        "name": "Day Planner: Re-sync tasks"
    },
    {
        "id": "obsidian-day-planner:show-day-planner-timeline",
        "name": "Day Planner: Show timeline"
    },
    {
        "id": "obsidian-day-planner:show-weekly-view",
        "name": "Day Planner: Show week planner"
    },
    {
        "id": "obsidian-day-planner:show-day-planner-today-note",
        "name": "Day Planner: Open today's Day Planner"
    },
    {
        "id": "obsidian-day-planner:insert-planner-heading-at-cursor",
        "name": "Day Planner: Insert Planner Heading at Cursor"
    },
```
Is there a solution to this?
in obsidian i get a template parsing error:
`Cannot read properties of undefined (reading 'id')`
Thanks

**Reply by Nudelsalad** [reply]
Nevermind, just noticed that it gets automatically linked :100: 
So the second line and the command script are not useful anymore i guess
